### PR TITLE
Reindex search resiliency

### PR DESCRIPTION
**Pull request by henningandersen**
Local reindex can now survive loosing data nodes that contain source
data. The original query will be restarted with a filter for
`_seq_no >= last_seq_no` when a failure is detected.

Part of #42612 and split out from #43187

A couple of follow-ups will be done (all indicated with todos) in addition to what is in the meta issue:

1. Fix the problem with `allowPartialSearchResult` in master
2. Build the `ScrollableHitSource` separately rather than in super constructor.
3. SeqNo mapping creation or avoid the issue.

### CODE REVIEW

**Comment by tbrooks8** [comment]
I will take a look at this tomorrow.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java [hunk 1]
```diff
@@ -135,7 +136,9 @@
         this.listener = listener;
         BackoffPolicy backoffPolicy = buildBackoffPolicy();
         bulkRetry = new Retry(BackoffPolicy.wrap(backoffPolicy, worker::countBulkRetry), threadPool);
-        scrollSource = buildScrollableResultSource(backoffPolicy);
+        // todo: this is trappy, since if a subclass override relies on subclass fields, they are not initialized. We should fix
```
Comment (tbrooks8) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java]: I agree. This was causing me problems in a recent PR.

**Review by Tim-Brooks — COMMENTED** [review]
This is mostly looking good to me. I don't see any issues with the approach or your design. It seems likely that the test failures are related to this change. Do you want to let me know when I should look again to officially approve?

**Comment by henningandersen** [comment]
Thanks @tbrooks8 and sorry for not fixing the test failure before now. It should be ready for another round (provided tests succeed).

**Comment by henningandersen** [comment]
@elasticmachine run elasticsearch-ci/1

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Pull request merged**